### PR TITLE
refactor: remove standard hook from utils.c 

### PR DESCRIPTION
**Pull request by steve-chavez**
Step towards solving https://github.com/supabase/supautils/issues/79.

### CODE REVIEW

File: src/supautils.c [hunk 1]
```diff
@@ -295,36 +297,31 @@ supautils_hook(PROCESS_UTILITY_PARAMS)
 				confirm_reserved_roles(get_rolespec_name(stmt->role), false);
 
 				if (!is_privileged_role()){
+					bypassrls_is_allowed = false;
```
Comment (soedirgo) [inline src/supautils.c:300]: does this do anything?

File: src/supautils.c [hunk 2]
```diff
@@ -295,36 +297,31 @@ supautils_hook(PROCESS_UTILITY_PARAMS)
...
-
-					if (strcmp(defel->defname, "bypassrls") == 0) {
-						if (bypassrls_option != NULL) {
-							ereport(ERROR,
-									(errcode(ERRCODE_SYNTAX_ERROR),
-									 errmsg("conflicting or redundant options")));
-						}
-						bypassrls_option = defel;
-					}
+					stmt_has_bypassrls = strcmp(defel->defname, "bypassrls") == 0;
```
Comment (soedirgo) [inline src/supautils.c:308]: need to exit the loop once a `bypassrls` DefElem is found

File: src/supautils.c [hunk 3]
```diff
@@ -295,36 +297,31 @@ supautils_hook(PROCESS_UTILITY_PARAMS)
...
 				}
 
-				// Defer setting bypassrls attribute if using `privileged_role`.
-				if (bypassrls_option != NULL) {
-					stmt->options = list_delete_ptr(stmt->options, bypassrls_option);
-				}
+				/* Allow doing BYPASSRLS */
+				if(bypassrls_is_allowed && stmt_has_bypassrls){
+					switch_to_superuser(privileged_extensions_superuser, &already_switched_to_superuser);
```
Comment (soedirgo) [inline src/supautils.c:313]: This allows for privilege escalation for the privileged_role, e.g. `alter role r superuser bypassrls;`

File: src/supautils.c [hunk 4]
```diff
@@ -414,15 +411,8 @@ supautils_hook(PROCESS_UTILITY_PARAMS)
...
-						//
-						// Duplicate/conflicting attributes will be caught by
-						// the standard process utility hook, so we can assume
-						// there's at most one bypassrls DefElem.
-						if (bypassrls_is_allowed && strcmp(defel->defname, "bypassrls") == 0) {
-							stmt_has_bypassrls = intVal(defel->arg) != 0;
-							intVal(defel->arg) = 0;
-						}
+						// Check if statement has BYPASSRLS
+						stmt_has_bypassrls = strcmp(defel->defname, "bypassrls") == 0;
```
Comment (soedirgo) [inline src/supautils.c:415]: need to do the assignment conditionally on the `strcmp`, like above

File: src/supautils.c [hunk 5]
```diff
@@ -452,15 +442,15 @@ supautils_hook(PROCESS_UTILITY_PARAMS)
 					// deferred in the original CREATE ROLE - the actual setting
 					// is done here.
 					if (bypassrls_is_allowed && stmt_has_bypassrls) {
-						Node *true_node = (Node *)makeInteger(true);
-						DefElem *bypassrls_option = makeDefElem("bypassrls", true_node, -1);
+						bool already_switched_to_superuser = false;
 
-						run_process_utility_hook(prev_hook);
+						switch_to_superuser(privileged_extensions_superuser, &already_switched_to_superuser);
```
Comment (soedirgo) [inline src/supautils.c:447]: privesc

File: src/supautils.c [hunk 6]
```diff
@@ -414,14 +413,13 @@ supautils_hook(PROCESS_UTILITY_PARAMS)
...
-						if (bypassrls_is_allowed && strcmp(defel->defname, "bypassrls") == 0) {
-							stmt_has_bypassrls = intVal(defel->arg) != 0;
-							intVal(defel->arg) = 0;
+						// Setting the superuser attribute is not allowed.
+						if (strcmp(defel->defname, "superuser") == 0) {
+							ereport(ERROR,
+									(errcode(ERRCODE_INSUFFICIENT_PRIVILEGE),
+									 errmsg("permission denied to alter role"),
+									 errdetail("Only roles with the %s attribute may alter roles with the %s attribute.",
+											   "SUPERUSER", "SUPERUSER")));
```
Comment (steve-chavez) [inline src/supautils.c:422]: @soedirgo Is this new? Would we need a new test for it? Assuming it's for `ALTER ROLE x SUPERUSER`.

Comment (soedirgo) [reply src/supautils.c:422]: Yup, will add a test for it

**Review by soedirgo — APPROVED** [review]

**Pull request merged**